src/main/java/com/mongodb/spark/sql/connector/schema/InferSchema.java

```diff
@@ -65,22 +69,42 @@ public final class InferSchema {
   public static final Metadata INFERRED_METADATA = Metadata.fromJson("{\"inferred\": true}");
 
   /**
-   * Infer the schema for the collection
+   * Infer the schema for the specified configuration.
    *
    * @param options the configuration options to determine the namespace to determine the schema for
    * @return the schema
    */
   public static StructType inferSchema(final CaseInsensitiveStringMap options) {
     ReadConfig readConfig = MongoConfig.readConfig(options.asCaseSensitiveMap())
         .withOptions(options.asCaseSensitiveMap());
-    ArrayList<Bson> samplePipeline = new ArrayList<>(readConfig.getAggregationPipeline());
-    samplePipeline.add(Aggregates.sample(readConfig.getInferSchemaSampleSize()));
-    return inferSchema(
-        readConfig.withCollection(coll -> coll.aggregate(samplePipeline)
+    CollectionsConfig collectionsConfig = readConfig.getCollectionsConfig();
+    return readConfig.withClient(client -> {
+      MongoDatabase database = client.getDatabase(readConfig.getDatabaseName());
+      Collection<String> collectionNames;
+      switch (collectionsConfig.getType()) {
+        case SINGLE:
+        case MULTIPLE:
+          collectionNames = collectionsConfig.getNames();
+          break;
+        case ALL:
+          collectionNames = database.listCollectionNames().into(new ArrayList<>());
+          break;
+        default:
+          throw fail();
+      }
+      List<BsonDocument> sampleDocuments = new ArrayList<>();
+      ArrayList<Bson> samplePipeline = new ArrayList<>(readConfig.getAggregationPipeline());
+      samplePipeline.add(Aggregates.sample(readConfig.getInferSchemaSampleSize()));
+      for (String collectionName : collectionNames) {
+        database
+            .getCollection(collectionName, BsonDocument.class)
+            .aggregate(samplePipeline)
             .allowDiskUse(readConfig.getAggregationAllowDiskUse())
             .comment(readConfig.getComment())
-            .into(new ArrayList<>())),
-        readConfig);
+            .into(sampleDocuments);
```

Comment:
We are sampling all the collections sequentially. If there are many collections, this may take significant amount of time. If users get back to us pointing that this is a problem, then we may think about possible solutions and implement them.

Reply:
That was my original approach, but Prakul explicitly asked for the schema inference, and said that it is an important feature.

Reply:
Fair enough - not sold its going to be practical, but I suppose for those that need multiple collection streaming it could be 🤷

Reply:
Slightly a noob question - can the user still add the schema definition explicitly (instead of auto-inference), if they chose to, when using multiple collections? 
We should be calling out in the documentation very clearly the suggestion to set schema explicitly, otherwise bear the performance implications. (We get this feedback consistently from user that they don't want to specify explicit schema as it tends to change over time, etc.)

Reply:
Yes, a user can specify an explicit schema as described in https://www.mongodb.com/docs/spark-connector/current/streaming-mode/streaming-read/.